session.php

```diff
@@ -122,24 +119,106 @@
             $pointid = $nextpointid;
         } else {
             // No more points to process - redirect to the session summary/submission screen.
-            redirect(new moodle_url('sessionsummary.php', ['sessionid' => $sessionid]), get_string('responsesaved', 'observation'),
-                null, \core\output\notification::NOTIFY_SUCCESS);
+            redirect(
+                new moodle_url('sessionsummary.php', ['sessionid' => $sessionid]),
+                get_string('responsesaved', 'observation'),
+                null,
+                \core\output\notification::NOTIFY_SUCCESS
+            );
             return;
         }
     }
 
-    redirect(new moodle_url('session.php', ['sessionid' => $sessionid, 'pointid' => $pointid]),
-        get_string('responsesaved', 'observation'), null, \core\output\notification::NOTIFY_SUCCESS);
+    redirect(
+        new moodle_url('session.php', ['sessionid' => $sessionid, 'pointid' => $pointid]),
+        get_string('responsesaved', 'observation'),
+        null,
+        \core\output\notification::NOTIFY_SUCCESS
+    );
     return;
 }
 
 // Render page.
-$PAGE->set_url(new moodle_url('/mod/observation/session.php', array('sessionid' => $sessionid)));
-$PAGE->set_title($course->shortname.': '.$observation->name);
+$pageurl = new moodle_url('/mod/observation/session.php', array(
+    'sessionid' => $sessionid, 'pointid' => $pointid,
+    'confirm' => $confirm, 'confirmsubmit' => $confirmsubmit
+));
+$pageurl->out(false);
+$PAGE->set_url($pageurl);
+$PAGE->set_title($course->shortname . ': ' . $observation->name);
 $PAGE->set_heading($course->fullname);
+
 echo $OUTPUT->header();
 echo $OUTPUT->heading(get_string('markingobservation', 'observation'), 2);
 
+if ($markingform->no_submit_button_pressed()) {
+    $fromform = $markingform->get_submitted_data();
+
+    // Cancel / abandon observation button pressed.
+    if (!is_null($fromform->abandonbutton)) {
+
+        // If no confirmation yet, display confirmation dialog.
+        if ($confirm === null) {
+            echo $OUTPUT->confirm(
+                get_string('confirmcancel', 'observation'),
+                new moodle_url($pageurl, ['confirm' => true]),
+                new moodle_url($pageurl, ['confirm' => false])
+            );
+        }
+    }
+
+    // Submit observation button pressed.
+    if (!is_null($fromform->submitobservation)) {
+
+        // If no confirmation yet, display confirmation dialog.
+        if ($confirmsubmit === null) {
+            echo $OUTPUT->confirm(
+                get_string('confirmsubmit', 'observation'),
+                new moodle_url($pageurl, ['confirmsubmit' => true]),
+                new moodle_url($pageurl, ['confirmsubmit' => false])
+            );
+        }
+    }
+    return;
+}
+
+if ($confirm !== null & $confirmsubmit !== 1) {
```

Comment:
Could this be `if ($confirm && !$confirmsubmit) ` or is there type coercion stuff that needs to be coded around